### PR TITLE
lock, windows: O_CREAT infers only GENERIC_WRITE

### LINKED ISSUE #3968: [Windows] mc cp fails with "Failed to copy ‘filename’. Insufficient permissions to access this file"
## Expected Behavior

`mc cp` should work.

## Current Behavior

`mc cp` fails with "Failed to copy ‘filename’. Insufficient permissions to access this file"

## Possible Solution

NA 
## Steps to Reproduce (for bugs)

1. mc cp \path\to\file myminio/bucket/obj
2. 
3.
4.

Stack trace on server
```
time="2017-03-24T14:23:33+02:00" level=error msg="Unable to create an object." cause="Prefix access is denied: bucketname/filename" source="[object-handlers.go:479:objectAPIHandlers.PutObjectHandler()]" stack="/home/builder/mygo/src/github.com/minio/minio/cmd/fs-v1.go:577:fsObjects.PutObject <autogenerated>:275:(*fsObjects).PutObject /home/builder/mygo/src/github.com/minio/minio/cmd/object-handlers.go:476:objectAPIHandlers.PutObjectHandler /home/builder/mygo/src/github.com/minio/minio/cmd/api-router.go:60:(objectAPIHandlers).PutObjectHandler-fm"
```

Minio Version
```
Version: 2017-02-16T01:47:30Z
Release-Tag: RELEASE.2017-02-16T01-47-30Z
Commit-ID: 3d98311d9f4ceb78dba996dcdc0751253241e697
```

## Your Environment

* Version used:
* Environment name and version (e.g. nginx 1.9.1):
* Server type and version:
* Operating System and version:
* Link to your project:

### PR DESCRIPTION
**Pull request by vadmeste**
## Description
Avoid using GENERIC_ALL flag when creating a file since it could
return permission denied in some Windows versions.

## Motivation and Context
Fixes #3968 

## How Has This Been Tested?
go test

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by vadmeste** [comment]
> Also can you print the relevant windows docs providing details on why GENERIC_ALL is an issue ?

GENERIC_ALL means read + write + execute, we don't really need execute in the first place that's why I didn't bother doing any investigation.

> You don't need to remove this part to avoid the perm denied.

TestLockFail() in lock_test.go doesn't like O_APPEND, so I made lock_windows.go follow lock_nix.go, let's discuss internally to validate this.

File: pkg/lock/lock_windows.go [hunk 1]
```diff
@@ -94,42 +88,26 @@ func open(path string, flag int, perm os.FileMode) (*os.File, error) {
 		access = syscall.GENERIC_READ
 	case syscall.O_WRONLY:
 		access = syscall.GENERIC_WRITE
-	case syscall.O_RDWR:
+	case syscall.O_RDWR,
```
Comment (harshavardhana) [inline pkg/lock/lock_windows.go]: Use a `fallthrough` don't put all of them in the same line.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by harshavardhana — APPROVED** [review]

**Comment by rawmaze** [comment]
Is there a Windows executable file produced from master which I can get to test if the problem has been fixed on my machine?

**Comment by vadmeste** [comment]
> Is there a Windows executable file produced from master which I can get to test if the problem has been fixed on my machine?

@imosk, no. You will need to install go on Windows and run `go get -u github.com/minio/minio`, we are available in the slack channel (https://minio.slack.com) to assist if you want.

**Comment by harshavardhana** [comment]
> Is there a Windows executable file produced from master which I can get to test if the problem has been fixed on my machine?

You could also use `docker pull minio/minio:edge` on windows.

**Pull request merged**